lucet-runtime/lucet-runtime-internals/src/context/context_asm.S

```diff
@@ -66,9 +74,27 @@ _lucet_context_backstop:
     // it happens to be an available pointer we can hang the contexts to swap off of.
     mov (%rbp), %rdi  /* parent context to arg 1 */
     mov 8(%rbp), %rsi  /* own context to arg 2 */
-    mov %rax, (8*8 + 8*16 + 8*0)(%rdi) /* store return values before swapping back -- offset is offsetof(struct lucet_context, retvals) */
-    mov %rdx, (8*8 + 8*16 + 8*1)(%rdi)
-    movdqu %xmm0, (8*8 + 8*16 + 8*2)(%rdi) /* floating-point return value */
+    mov %rax, (10*8 + 8*16 + 8*0)(%rdi) /* store return values before swapping back -- offset is offsetof(struct lucet_context, retvals) */
+    mov %rdx, (10*8 + 8*16 + 8*1)(%rdi)
+    movdqu %xmm0, (10*8 + 8*16 + 8*2)(%rdi) /* floating-point return value */
+
+    // rax has been stored above already, so we can reuse it now to load the 
+    // pointer to the `terminable` flag.
+    mov 16(%rbp), %rax
+    mov $0, %cl
+    // there may be standing executors that will try to execute this instance,
+    // take the `terminable` flag so they will fail.
+    // x86 xchg has an implicit "lock" semantic, so we don't need the prefix here.
+    xchgb (%rax), %cl
+    test %cl, %cl
+
+    // if we got 0 from the `xchg`, some terminator beat us in trying to
+    // stop this instance, and it's not safe to start swapping contexts.
+    //
+    // so, just hang out for a bit and wait for the terminator to do its thing
+_hang:
+    jz _hang
```

Comment:
It is, but this might deserve some clarification as to _why_ - if the zero flag is set, `cl` is zero, and the attempt to acquire the flag failed. In that case, we're about to exit, but something else intends to terminate the guest, so we do spin lest we enter a critical section (context swap) or are outright in host code.

In the case we got the flag, `cl` will be `1`, the zero flag will be unset, and we fall right through the loop, since we acquired the termination flag and nothing else will try stopping in a subsequent critical section.

I was going to write out what this looks like less golf'd but it turns out to be clearer anyway:
```
; test if we actually acquired the flag
test %cl, %cl
jnz lucet_context_swap
; we didn't get the flag, so something else will be terminating this guest
_await_termination:
  jmp _await_termination

Reply:
Got it, that makes sense and the new version makes that a lot more evident